`panic` when `postpackage` hook fails

```
$ azd version
azd version 0.7.0-beta.2-daily.2671790 (commit 23cf941b3e934e2f51606d467041e892d49e65bd)
```

Reproduce:

1. `azd init` and pick your favorite template
2. Add a hook that always fails. Example below:
```yaml
  api:
    project: ./src/api
    language: csharp
    host: appservice
    hooks:
      postpackage:
        shell: pwsh
        run: Write-Error "failing"; exit 1
        continueOnError: false
        interactive: false
```
3. Run `azd package`


Expected: hook fails with error "postpackage failed"
Actual: hook fails with the following error:

```
panic: Task result has already been set! Task cannot have both a result and an error.

goroutine 87 [running]:
github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskContext[...]).SetError(...)
        D:/a/_work/1/s/cli/azd/pkg/async/task_context.go:23
github.com/azure/azure-dev/cli/azd/pkg/project.(*serviceManager).Package.func1(0xc0000f0c90)
        D:/a/_work/1/s/cli/azd/pkg/project/service_manager.go:356 +0x6fe
github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run.func1()
        D:/a/_work/1/s/cli/azd/pkg/async/task.go:144 +0x11d
created by github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run
        D:/a/_work/1/s/cli/azd/pkg/async/task.go:140 +0x1cf
```

![image](https://user-images.githubusercontent.com/2322434/229606833-39d069b4-f178-4451-afb5-13df32852f52.png)

--- a/cli/azd/pkg/project/service_manager.go
+++ b/cli/azd/pkg/project/service_manager.go
@@ -329,6 +329,8 @@ func (sm *serviceManager) Package(
 			buildOutput.Restore = restoreResult
 		}
 
+		var packageResult *ServicePackageResult
+
 		err = serviceConfig.Invoke(ctx, ServiceEventPackage, eventArgs, func() error {
 			frameworkPackageTask := frameworkService.Package(ctx, serviceConfig, buildOutput)
 			syncProgress(task, frameworkPackageTask.Progress())
@@ -346,8 +348,8 @@ func (sm *serviceManager) Package(
 				return err
 			}
 
-			task.SetResult(serviceTargetPackageResult)
-			sm.setOperationResult(ctx, serviceConfig, string(ServiceEventPackage), serviceTargetPackageResult)
+			packageResult = serviceTargetPackageResult
+			sm.setOperationResult(ctx, serviceConfig, string(ServiceEventPackage), packageResult)
 
 			return nil
 		})
@@ -356,6 +358,8 @@ func (sm *serviceManager) Package(
 			task.SetError(fmt.Errorf("failed packaging service '%s': %w", serviceConfig.Name, err))
 			return
 		}
+
+		task.SetResult(packageResult)
 	})
 }
 

--- a/cli/azd/pkg/project/service_manager_test.go
+++ b/cli/azd/pkg/project/service_manager_test.go
@@ -2,6 +2,7 @@ package project
 
 import (
 	"context"
+	"errors"
 	"strings"
 	"testing"
 
@@ -10,6 +11,7 @@ import (
 	"github.com/azure/azure-dev/cli/azd/pkg/convert"
 	"github.com/azure/azure-dev/cli/azd/pkg/environment"
 	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/ext"
 	"github.com/azure/azure-dev/cli/azd/pkg/infra"
 	"github.com/azure/azure-dev/cli/azd/pkg/ioc"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools"
@@ -256,6 +258,73 @@ func Test_ServiceManager_CacheResults(t *testing.T) {
 
 	require.False(t, *buildCalled)
 	require.Same(t, buildResult1, buildResult2)
+}
+
+func Test_ServiceManager_Events_With_Errors(t *testing.T) {
+	tests := []struct {
+		name      string
+		eventName string
+		run       func(ctx context.Context, serviceManager ServiceManager, serviceConfig *ServiceConfig) (any, error)
+	}{
+		{
+			name: "restore",
+			run: func(ctx context.Context, serviceManager ServiceManager, serviceConfig *ServiceConfig) (any, error) {
+				restoreTask := serviceManager.Restore(ctx, serviceConfig)
+				logProgress(restoreTask)
+				return restoreTask.Await()
+			},
+		},
+		{
+			name: "build",
+			run: func(ctx context.Context, serviceManager ServiceManager, serviceConfig *ServiceConfig) (any, error) {
+				buildTask := serviceManager.Build(ctx, serviceConfig, nil)
+				logProgress(buildTask)
+				return buildTask.Await()
+			},
+		},
+		{
+			name: "package",
+			run: func(ctx context.Context, serviceManager ServiceManager, serviceConfig *ServiceConfig) (any, error) {
+				packageTask := serviceManager.Package(ctx, serviceConfig, nil)
+				logProgress(packageTask)
+				return packageTask.Await()
+			},
+		},
+		{
+			name: "deploy",
+			run: func(ctx context.Context, serviceManager ServiceManager, serviceConfig *ServiceConfig) (any, error) {
+				deployTask := serviceManager.Deploy(ctx, serviceConfig, nil)
+				logProgress(deployTask)
+				return deployTask.Await()
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			mockContext := mocks.NewMockContext(context.Background())
+			setupMocksForServiceManager(mockContext)
+			env := environment.EphemeralWithValues("test", map[string]string{
+				environment.SubscriptionIdEnvVarName: "SUBSCRIPTION_ID",
+			})
+			sm := createServiceManager(mockContext, env)
+			serviceConfig := createTestServiceConfig("./src/api", ServiceTargetFake, ServiceLanguageFake)
+
+			eventTypes := []string{"pre", "post"}
+			for _, eventType := range eventTypes {
+				t.Run(test.eventName, func(t *testing.T) {
+					test.eventName = eventType + test.name
+					_ = serviceConfig.AddHandler(ext.Event(test.eventName), func(ctx context.Context, args ServiceLifecycleEventArgs) error {
+						return errors.New("error")
+					})
+
+					result, err := test.run(*mockContext.Context, sm, serviceConfig)
+					require.Error(t, err)
+					require.Nil(t, result)
+				})
+			}
+		})
+	}
 }
 
 func setupMocksForServiceManager(mockContext *mocks.MockContext) {

--- a/cli/azd/pkg/project/service_manager_test.go
+++ b/cli/azd/pkg/project/service_manager_test.go
@@ -314,9 +314,12 @@ func Test_ServiceManager_Events_With_Errors(t *testing.T) {
 			for _, eventType := range eventTypes {
 				t.Run(test.eventName, func(t *testing.T) {
 					test.eventName = eventType + test.name
-					_ = serviceConfig.AddHandler(ext.Event(test.eventName), func(ctx context.Context, args ServiceLifecycleEventArgs) error {
-						return errors.New("error")
-					})
+					_ = serviceConfig.AddHandler(
+						ext.Event(test.eventName),
+						func(ctx context.Context, args ServiceLifecycleEventArgs) error {
+							return errors.New("error")
+						},
+					)
 
 					result, err := test.run(*mockContext.Context, sm, serviceConfig)
 					require.Error(t, err)

